File: 304. Range Sum Query 2D - Immutable.py
class NumMatrix(object):
    def __init__(self, matrix):
        """
        initialize your data structure here.
        :type matrix: List[List[int]]
        """
        m=len(matrix)
        if m>0:
            n=len(matrix[0])
            self.T=[[0 for j in xrange(n)] for i in xrange(m)]
            self.T[0][0]=matrix[0][0]
            for i in xrange(1,m):
                self.T[i][0]=self.T[i-1][0]+matrix[i][0]
            for j in xrange(1,n):
                self.T[0][j]=self.T[0][j-1]+matrix[0][j]
            for i in xrange(1,m):
                for j in xrange(1,n):
                    self.T[i][j]=matrix[i][j]+self.T[i][j-1]+self.T[i-1][j]-self.T[i-1][j-1]
            print self.T

    def sumRegion(self, row1, col1, row2, col2):
        """
        sum of elements matrix[(row1,col1)..(row2,col2)], inclusive.
        :type row1: int
        :type col1: int
        :type row2: int
        :type col2: int
        :rtype: int
        """
        if row1==0 and col1==0:
            return self.T[row2][col2]
        elif row1>0 and col1==0:
            return self.T[row2][col2]-self.T[row1-1][col2]
        elif row1==0 and col1>0:
            return self.T[row2][col2]-self.T[row2][col1-1]
        return self.T[row2][col2]-self.T[row1-1][col2]-self.T[row2][col1-1]+self.T[row1-1][col1-1]
        


# Your NumMatrix object will be instantiated and called as such:
# numMatrix = NumMatrix(matrix)
# numMatrix.sumRegion(0, 1, 2, 3)
# numMatrix.sumRegion(1, 2, 3, 4)

File: 304. Range Sum Query 2D - Immutable.cpp
class NumMatrix {
public:
    vector <vector<int>> T;
    NumMatrix(vector<vector<int>> &matrix) {
        int m = matrix.size();
        if(matrix.empty()) return;
        int n = matrix[0].size();
        T = vector<vector<int>> (m, vector<int>(n));
        for(int i = 0; i < m; i++)
        {
            for(int j = 0; j < n; j++)
            {
                if(i == 0 && j == 0)
                    T[i][j] = matrix[0][0];
                else if(i == 0)
                    T[i][j] = T[i][j-1] + matrix[i][j];
                else if(j == 0)
                    T[i][j] = T[i-1][j] + matrix[i][j];
                else
                    T[i][j] = T[i-1][j] + T[i][j-1] + matrix[i][j] - T[i-1][j-1];
            }
        }
    }

    int sumRegion(int row1, int col1, int row2, int col2) {
        if (row1 == 0 && col1 == 0)
            return T[row2][col2];
        else if (row1 > 0 && col1 == 0)
            return T[row2][col2] - T[row1-1][col2];
        else if (row1 == 0 && col1 > 0)
            return T[row2][col2] - T[row2][col1-1];
        return T[row2][col2] - T[row1-1][col2] - T[row2][col1-1] + T[row1-1][col1-1];
    }
};


// Your NumMatrix object will be instantiated and called as such:
// NumMatrix numMatrix(matrix);
// numMatrix.sumRegion(0, 1, 2, 3);
// numMatrix.sumRegion(1, 2, 3, 4);
